### DISCUSSION TITLE
How to directly create a user

### DISCUSSION THREAD

**Discussion by ProjectMoon** [supabase/supabase#2346] [Questions]
Is there a way to directly create a user, without going through sign up flow, or the invite flow from the admin dashboard?

**Comment by burggraf** [comment]
I'm not 100% sure I understand what you mean by `"directly create a user"` but  users are simply stored in the table `auth.users` so you could always write a `PostgreSQL function` to create those records manually.  I would just be careful, though, that you create the records with all the required fields so they look the same as the records that are created using the sign up flow.

**Reply by ProjectMoon** [reply]
Skipping the sign-up process, or creating a user without an email address.

Somewhat unrelated, I'm not getting invite emails from supabase, or they arrive spottily. This is a Google Business/Google Workspace or whatever it's called these days account. One user was able to get the email invite, another was not. Should I direct that to a support ticket or something?

**Reply by burggraf** [reply]
You can open up a bug report if you can maybe narrow the issue down a bit.  Try to figure out when or in what cases it's failing (as it works fine for me in all my apps.).  Start by investigating the user who didn't get the email.  What's the domain of their email address, did it show up in their spam or junk folder, etc?  Thanks! 

And regarding skipping the sign-up process or creating a user without an email address -- what I said should work, just create entries in `auth.users`.

**Reply by ProjectMoon** [reply]
The domain of the email is just one of my personal domains. But they do not get the email at all for some reason. It just never shows up, in either spam or inbox. Other emails to and from the user work though. I've deleted and created the user several times in the Supabase admin UI, which I assume would send an invitation email each time.

**Reply by burggraf** [reply]
Do you have any tools available for the domain to see what's happening to the mails, to see if they're being rejected for some reason?  I guess it depends on who your mail provider is.

**Reply by richcorbs** [reply]
Was this ever solved? I'm trying to do a similar thing but cannot get it working yet:

0. I disabled signup.
1. I create the record in public.users.
2. A trigger function creates the corresponding record in auth.users.
3. When I submit just the new user's email for magic link login it returns "Database error finding user".

In the auth.users table the only difference that I can see is that the encrypted_password and confirmation_token fields are not exactly the same format (mine are UUID and supabase's are hashes).

My reasons for this flow are:

1. I do not want signups.
2. I do not want to have to create the new auth.users via the dashboard.
3. I want the teams to be able to add their own public.users via the UI and have it work.

I tried both security models (creator vs caller) on my trigger function and neither worked. I thought for sure it was a permissions thing and that this would fix it. It didn't.

**Reply by burggraf** [reply]
Can you just call `auth.signUp()` [link](https://supabase.io/docs/reference/javascript/auth-signup) from an admin page, so the users don't sign up themselves, but your admin teams can enter users themselves?  So you'd create a form and a button and they could enter users themselves?

**Reply by richcorbs** [reply]
That would require signup being enabled. Right?

If signup is enabled then anyone who goes to the login page and enters an email address would be signed up. That's not what I need.

**Reply by burggraf** [reply]
They wouldn't be signed up unless you called `auth.signUp()` from the login page.  I'm not suggesting you do that.  The login page would only allow people to log in (there would be no "sign up" button on the page.)

I'm suggesting a separate page that only teams (admin users) or whoever you want to give the ability to create users would go to to create the users.  A "Create Users" page, per se.  Make sense?

**Reply by richcorbs** [reply]
Yep, I understand what you are saying.

Even with signup disabled, calling `supabase.auth.signIn` with just an email creates an `auth.users` record. Unless there is a way to turn that off I don't think that helps. Maybe that's a bug?

Also, if signup is disabled, how would we get new auth.users outside of the dashboard? Seems like a viable use case.

**Reply by burggraf** [reply]
I don't think you need to disable signups for this use case.  Just don't make a "Sign Up" button in your application.  If there's no page or form for a user to sign up, you should be fine, right?  So in essence, you're just putting the sign up process in the hands of a select few people who do the signups.  Unless I'm missing your intent here?

**Reply by richcorbs** [reply]
That is my intent.

I don't think we're communicating. :-) I'll try to be more clear.

Calling `signIn` with an email creates the auth.users record and effectively allows anyone to sign up. That is NOT what I'm trying to accomplish.

Maybe I don't understand what "signIn" means. To me it should mean: only sign me in if I'm an existing user. It doesn't mean "create a user account for me as if I were signing up and allow me to signin".

I would be happy to use the signUp method as you describe if signIn didn't create a new auth.users record enabling the equivalent of signUp.

Am I making any more sense?

**Reply by ghost** [reply]
> I'm not 100% sure I understand what you mean by `"directly create a user"` but users are simply stored in the table `auth.users` so you could always write a `PostgreSQL function` to create those records manually. I would just be careful, though, that you create the records with all the required fields so they look the same as the records that are created using the sign up flow.

@burggraf Can you provide pointers about how to properly generate the encrypted password when inserting rows into that table? More specifically, what is the encryption algorithm and where can I find the secret that is fed into the algo (is it the service_role secret key)?

**Reply by burggraf** [reply]
Sorry I've never done that, you might have to ask this over in the [gotrue repo](https://github.com/supabase/gotrue).  What I've been doing is creating my own auth links and sending them directly via my own email provider:  [https://github.com/burggraf/supabase-roll-your-own-auth](https://github.com/burggraf/supabase-roll-your-own-auth)

**Reply by ghost** [reply]
Thanks a lot, @burggraf! Pointing me to the GoTrue repo was helpful - I saw that the password was only hashed (i.e. not actually encrypted with a secret) using bcrypt:

https://github.com/supabase/gotrue/blob/70ffa779d7f1a58e5463a513289597f3abc9d8e6/models/user.go#L242

This makes creating new users by directly inserting them in the table quite easy. One note for everyone else reading this - as @burggraf  mentioned above, getting all the fields in the row right might be tricky so you can probably just copy the values from another existing row and only change the relevant ones (e.g. id, email, encrypted password).

**Comment by burggraf** [comment]
`Calling signIn with an email creates the auth.users record and effectively allows anyone to sign up. That is NOT what I'm trying to accomplish.`

No.  Calling `signIn` signs a user in.  Calling `signUp` creates a new user.  `signIn` only signs you in if you're already a user in the `auth.users` table.  If you're not already a user, you get an error (user not found, or something like that).

**Comment by burggraf** [comment]
I guess, though, according to the docs, calling `signIn` without a password sends a magic link.  That should be able to be disabled, I think?

**Comment by richcorbs** [comment]
@burggraf I appreciate that you are trying to help but this is not what I see in real life.

1. I call `supabase.auth.signIn({email: 'rich@gmail.com'})`.
2. In the network tab I see it hit this path `/auth/v1/magiclink`.
3. When I go to the dashboard > Authentication > Users there is a new record `waiting for verification` for `rich@gmail.com`.

We overlapped comments. Yeah, feels like a bug somewhere, somehow. Signin/magiclink shouldn't EVER create an account IMO.

**Reply by burggraf** [reply]
Can you always call `signIn` with a password parameter?  This should not create a new user, I think.

**Reply by burggraf** [reply]
You might want to create an issue with a feature request, asking for the ability to turn off Magic Links.

**Reply by richcorbs** [reply]
Not seeing how always calling `signIn` with a password would help. What are you thinking?

**Reply by richcorbs** [reply]
Which repo would be best for an issue? supabase/supabase?

**Reply by burggraf** [reply]
If you call signIn with an email and a password parameter, a new user should not be created.  You should just get an error if the user doesn't exist.

**Reply by richcorbs** [reply]
Correct I get back `invalid credentials`.

**Reply by burggraf** [reply]
I get "invalid login credentials" if I pass an invalid email / password combination.  No user is created.

**Reply by burggraf** [reply]
LOL we crossed paths.

**Reply by richcorbs** [reply]
Still not seeing how that moves me forward. I want magiclinks to work when the account exists. I don't want magiclinks to create an account when one doesn't exist.

**Reply by burggraf** [reply]
Ah, that makes more sense now.  So the feature request would be to only allow magic links for existing accounts.  That's never come up before.  Sort of a "pre-qualified user list".

**Reply by burggraf** [reply]
You could hack it maybe by writing a PostgreSQL function that hits the signIn endpoint of your auth system, then just have the function check for the user in `auth.users` before it calls the `signIn` to send the magiclink.

So user enters their email, you pass that to your function, your function looks for the user, and if it finds that they exist your function calls `signIn`.

**Reply by burggraf** [reply]
I'm doing something similar here:
https://github.com/burggraf/supabase-roll-your-own-auth/blob/main/02_generate_auth_link_function.sql

In a new lib I'm working on:  [supabase-roll-your-own-auth](https://github.com/burggraf/supabase-roll-your-own-auth)

**Reply by richcorbs** [reply]
Interesting idea. I'll try it out depending on how the issue goes.

https://github.com/supabase/supabase/issues/3033

**Reply by richcorbs** [reply]
Thanks for your time and insights.

**Reply by burggraf** [reply]
You would just need to call this endpoint from your function:  https://github.com/supabase/gotrue#post-magiclink-recommended-to-use-otp-instead-see-above

**Reply by richcorbs** [reply]
I got the "check_for_auth_user" function working. Thanks for the idea.

Now I can remove my other trigger and function that were trying to create auth.users records and just use the signUp functionality.

**Reply by burggraf** [reply]
Awesome!  Let us know how this works out.
